src/main/scala/linting/rule/LintAmbiguousModuleNames.scala

```diff
@@ -0,0 +1,331 @@
+  def removePortInfo(port: Port): Port = {
+    port.copy(info = NoInfo)
+  }
+
+  def removeModuleInfo(mod: Module): Module = {
+    mod.copy(
+      ports = mod.ports.map(removePortInfo(_)),
+      body = removeStatementInfo(mod.body),
+      info = NoInfo)
+  }
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass,
+    PortStructureNamingStrategy.getClass,
+    ContentStructureNamingStrategy.getClass,
+    ContentNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies a desired name for a module
+  */
+case class ModuleNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): ModuleNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** An example of how to customize the names of certain modules using an Aspect
+  */
+case object StabilizeNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    Select.collectDeep(top) {
+      // annotating all Queues with a more descriptive desired name
+      case m: Queue[_] => Seq(new ModuleNameAnnotation(s"Queue_${m.genType.getClass.getSimpleName}_entries_${m.entries}", m.toTarget))
+      case m: TLMonitor => Seq(new ModuleNameAnnotation(m.desiredName, m.toTarget))
+
+      // annotating specific instances
+      case th: TestHarness =>
+        val core = th.ldut.rocketTiles.head.module.core
+        Seq(
+          ModuleNameAnnotation("Rocket_Core_0", core.toTarget),
+          NamingStrategyAnnotation(ExactNamingStrategy, core.toTarget),
+        )
+    }.flatten.toSeq
+  }
+}
+
+/** This LintRule checks for module name collisions and optionally renames them to stable hash names
+  *
+  * Module name collisions occur when different [[Module]]s are annotated with
+  * [[ModuleNameAnnotation]]s that have the same desiredName. If one of the
+  * modules is targeted by [[NamingStrategyAnnotation]] then the transform will
+  * attempt to rename modules according to the naming strategy. If the
+  * conflicting modules cannot be disambiguated by the naming strategy then a
+  * [[LintViolation]] is emitted. [[ExactNamingStrategy]] can be used to
+  * enforce that there are not colliding module names.
+  *
+  * If no naming strategy is specified then a sequence of increasingly unstable
+  * [[NamingStrategy]]s is tried until all modules can be disamgiguated.
+  */
+final class LintAmbiguousModuleNames extends LintRule {
+  val recommendedFix: String = "override desiredName based on module parameters"
+
+  val lintName: String = "ambiguous-module-names"
+
+  private val strategyOrder = Seq(
```

Comment:
So if I apply this lint check, then automatically the renaming also applies? Or is the desired strategy applied separately from the lint check? I guess I'm confused by the overall user flow here. I would guess it would be separate options for:

1. Apply lint check at (some hierarchical level) that "desiredNames must not conflict".
2. Apply annotation (at some hierarchical level) "how to name things with strategies when desiredName conflicts"
3. For a specific module type / module instance, override desiredName after the fact (like your example for the TLMonitor and Queue).

I think 1, 2, and 3 should be able to be applied independently. Is my interpretation correct?
It doesn't make a ton of sense for 1 and 2 to be applied at the same time.

Reply:
Actually, I re-read your PR description and realized that my 1, 2, 3 do indeed correspond to 3 separate annotations, so I think I agree with the general approach. Still reading up on the details

Reply:
Well, not sure about this. Can I lint names to reject any conflicts without applying the naming strategies?

Reply:
yes, but I think some of the defaults are not correct. by default it will try to resolve name conflicts using the renaming strategies with the heuristic. if you want to just lint names you'll need to also specify `ExactNamingStrategy` which will reject any conflicts. but I think it should by default use `ExactNamingStrategy`?

Reply:
and we don't really need to use the strategy heuristic and should always be specifying a specific naming strategy to override it?